### PR TITLE
Load JSON $refs 

### PR DESCRIPTION
**Pull request by econchick**
Addresses #4; includes commits from #19 

ping @benhamill @nvillalva @jhl2343 if you'd like to review my approach.

### CODE REVIEW

File: docs/extendedusage.rst [hunk 1]
```diff
@@ -344,6 +351,279 @@ would be ``/foo/bar/{id}``, and the absolute URI path would be
...
+      "type": "object",
+      "properties": {
+        "album_type": {
+          "type": "string",
+          "description": "The type of the album: one of 'album', 'single', or 'compilation'."
+        },
+        "artists": {
+            "type": "array",
+            "description": "The artists of the album. Each artist object includes a link in href to more detailed information about the artist.",
+            "items": [{ "$ref": "artist.schema.json#properties" }]
```
Comment (benhamill) [inline docs/extendedusage.rst]: Should this be `"file:artist.schema.json#properties"`? It's a little late, so I might be fuzzy.

**Comment by benhamill** [comment]
This looks at least mostly-good to me. I want (before or after merge) to sit down with the tests when I'm less sleepy and really grok them and compare the inputs and expectations. With them broken across files like that, it's a bit hard to get, but I totally understand why you did it that way.

File: docs/extendedusage.rst (hunk 1 above)
Comment (econchick) [inline docs/extendedusage.rst]: There isn't a requirement for `file:` to prepend the filename, actually.

Comment (benhamill) [inline docs/extendedusage.rst]: Huh. So in the case of a missing schema, it needs to assume `file:` (relative)?

Comment (econchick) [inline docs/extendedusage.rst]: Sorry - I'm a bit confused - what do you mean by "in the case of a missing schema"?

**Comment by econchick** [comment]
A bit of a design question for anyone (@benhamill looking at you :smile: ) - the [RAML spec says](http://raml.org/spec.html#includes), in relation to a RAML file with `!includes`:

> If the original file is fetched as an HTTP resource, the included file SHOULD be fetched over HTTP.

So if a RAML file was on a remote server requested over HTTP, then included files should also be fetched over HTTP (versus HTTPS, or another protocol).

The JSON Schema [draft 3](https://tools.ietf.org/html/draft-zyp-json-schema-03#section-3.2) and [draft 4](https://tools.ietf.org/html/draft-zyp-json-schema-04#section-5.4) explicitly says the specification is agnostic to protocol.  So theoretically, a `$ref` could point to an FTP server, or something (it looks like this works with the `jsonref` library).  I may have missed it, but I can't find anything within the drafts that say if I get a remote JSON file over HTTP, any remote `$ref`s have to match protocols.

Q: If a RAML file is remote, should the `$ref` protocol match the protocol over which the RAML file was fetched?  e.g. if I got a RAML file over HTTPS, and there's a JSON `$ref` pointing to HTTP, or FTP, or anything other than HTTPS, should `ramlfications` error out?

File: docs/extendedusage.rst (hunk 1 above)
Comment (econchick) [inline docs/extendedusage.rst]: side note - I just included a test that makes sure `file:` does work.  So the following cases should all work (all with or without `#fragments`):

* `{"$ref": "local.schema.json"}`
* `{"$ref": "file:local.schema.json"}`
* `{"$ref": "include/local.schema.json"}`
* `{"$ref": "file:include/local.schema.json"}`
* `{"$ref": "/absolute/path/include/local.schema.json"}`
* `{"$ref": "file:/absolute/path/include/local.schema.json"}`

**Comment by econchick** [comment]
meh - I've concluded that it's not worth it (enforcing protocols for remote `$ref`s)

File: docs/extendedusage.rst (hunk 1 above)
Comment (benhamill) [inline docs/extendedusage.rst]: Sorry. Too many uses of "schema". I meant the protocol section of the URL.

And, FWIW, I think `file:/foo` is an invalid URI. I think it'd have to be `file:///foo`. If I'm recalling all the reading of specs I did about a month ago correctly, which, who knows.

Comment (econchick) [inline docs/extendedusage.rst]: /me facepalms

you're totally right.  I'll fix the patch in a bit.

Re: you're question about assuming a scheme.  Looking through the [jsonref](https://github.com/gazpachoking/jsonref/blob/master/jsonref.py) implementation, it uses `urllib2.urlopen` to handle the "getting" of the file, whether it's remote (HTTP/S), or local (with `file`).

The code proposed, specifically [this line](https://github.com/spotify/ramlfications/pull/40/files#diff-85519d2b28a6c19a3f995368d006d5a6R51), feeds in a base URI, which ends up being the directory of where the RAML file lives.  This base URI will always be a `file`  (which if `ramlfications` wants to support loading a RAML file from a remote URL, this will probably need a bit more logic).

`jsonref` has a lot of magic that - for me - is hard to follow.  There has to be some way that it _doesn't_ just build the `base_uri` + `$ref` so it ends up with something like `file:///foo/bar/https://example.com` because it's obvious that it doesn't happen.  But I'm done for the morning trying to understand it :)

Comment (benhamill) [inline docs/extendedusage.rst]: I don't blame you. IIRC, the various RFCs say that the string should be a URL (or a portion thereof, in the case of the `id` thing) and that means it _should_ have a protocol, right? Whether the code in question also requires it or not? I'm pulling this string in the interest of making the documentation examples as close to "ideal" as possible, so that if people learn from them without reading 4 RFCs (ugh), they'll still be writing good RAML/JSON Schema.

Comment (econchick) [inline docs/extendedusage.rst]: You're right w/r/t `id` should be a string and a URI if defined ([here's what it says in the draft 4 spec](https://tools.ietf.org/html/draft-zyp-json-schema-04#section-7.2)) - although it's not required.  

Relative `$ref`s [should be resolved](https://tools.ietf.org/html/draft-zyp-json-schema-03#section-5.28) against the URI of the current schema (the `id` of the closest parent).  If a relative `$ref` is defined with no `id` in the schema, then - if I read [this section](https://tools.ietf.org/html/draft-zyp-json-schema-03#section-5.27) correctly, you assume a relative `$ref` is relative to the URI of the parent (which I guess, if loaded locally, then it'd be the current directory).  FWIW the `jsonref` library _looks_ like it will take preference of a defined `id` over a provided `base_uri` parameter, which seems like the correct way given how the RFC for draft 3 reads (I should also write a test case for this).

Re: protocols - the RFCs only state **URIs**, which doesn't necessarily mean that protocol scheme is required (I honestly had to [look this up](https://danielmiessler.com/study/url_vs_uri/) - e.g.  `files.example.com` is a _URI_, `ftp://files.example.com` and `http://files.example.com` are _URLs_).  So the `id` URI does not need to have a protocol attached to it.  I'm not sure how `jsonref` handles undefined schemes (will also make a test case for this).

I do agree that the docs _should_ teach folks how to write good RAML/JSON (it's honestly how I learn best regarding best practices and such).  I do see some RAML-specific questions within the `#ramlfications` freenode channel (partly because `#raml` channel never has any RAML folks in it despite them owning the channel).  I will create an issue at some point today to help improve that aspect.

Not sure if I just went off on a tangent - I honestly spent an hour or so this morning rereading RFCs  and I may have gone done a rabbit hole :)  But this convo has definitely helped me to learn a lot!

Comment (benhamill) [inline docs/extendedusage.rst]: Ugh. URIs. URLs. So confusing. Tangent that may have been, but that was really interesting. Thanks for laying it all out.

**Comment by benhamill** [comment]
I read through some test cases and my head started to hurt. Realistically, I don't think I'm going to grok them the way I'd like. I say merge it.

**Comment by econchick** [comment]
Oooph @benhamill I don't like the sound of that! What's confusing? Is there something that could be clearer/better?

**Comment by benhamill** [comment]
Heh. There could be fewer permutations to test! ;) But seriously: I think it's just a complex problem space with a lot of corners and the data under test necessarily has to be spread across a number of files for each test and given the amount of hard thinking I'm doing in other aspects of my life, I can't marshal the mental resources to hold it all in my head at once. It's not the code; it's me.

**Comment by econchick** [comment]
closes #19 and #4

**Comment by benhamill** [comment]
Woo! :metal:
